src/mlir/design.md

```diff
@@ -0,0 +1,158 @@
+we get at this stage is already complete, and checked for syntactic and semantic
+errors.
+
+Lowering errors can still occur, if the semantics would be correct in some cases
+but not in this specific case, for example, when using it with a different
+runtime functionality.
+
+The MLIR dialect that we'll develop at this stage will be solely for representing
+the lowering concepts in order to simplify optimisations. In theory, the AST
+could have been directly lowered to LLVM IR, but the LLVM pipeline wouldn't be
+able, for example, to optimise based on type and region information.
+
+## Module Boundaries
+
+Some symbols are private to the module and don't escape its boundaries. If we
+lower all functions already mangled into the IR, we'd have to scan the whole
+module every time for the functions and types that belong to them, to run module
+local optimisations.
+
+For this reason, we have chosen to create an MLIR `module` for each class/module,
+including the root module. For example:
+```
+// Verona code in directory FooBar
+class Foo { foo() { ... } }
+
+// Equivalent MLIR structure
+module @__ {
+  module @FooBar {
+    module @Foo {
+      func @foo() {
+        ...
+      }
+    }
+  }
+}
+```
+
+This structure, however, does not convert naturally to LLVM dialect. So we have
+a later pass that mangles the names and moves all functions to the root module
+just before lowering to LLVM dialect. The example above would be:
+
+```
+module @__ {
+  func @____FooBar__Foo__foo() {
+    ...
+  }
+}
+```
+
+And all calls to that function would also be changed in every other function that
+calls this one.
+
+## Load/Store Semantics
+
+Stack variables are allocated by LLVM IR with an `alloca` instruction. MLIR has
+a similar concept and we use that in the first implementation. The main problem
+with this approach is that there is only one dialect that has alloca/load/store:
+`memref`.
+
+MemRefs represent the memory used by `tensors`, which are multi-dimensional
+objects (matrices) of the same type. Verona is obviously not restricted to
+tensors as types, so we'll need to extend this to structure types, with complex
+layout semantics (ex. union types). For that, we'll need a set of dialect
+instructions that can be later lowered directly into LLVM for allocating stack
+space on complex layout semantics.
+
+For now, every function argument and local declaration (`let`, `var`) are
+allocated in the stack and its value stored in the memory space. This simplifies
+code generation (all variable symbols point to an address), but it doesn't help
+with value symbols and are really slow.
+
+The LLVM pipeline, even at O1, recognises most of the argument, temporary and
+local allocas as bogus and optimises them away, so this isn't a critical issue
+from a performance point of view, but we will probably need to update the symbol
+table soon to account for more types and states of the variables we hold (ex.
+`let` vs `var`).
```

Comment:
Is there an MLIR equivalent of mem2reg / SROA?  This representation feels like it may make some MLIR-level optimisations harder but it might be fine.

Reply:
Not that I'm aware of, but the plan is to not need one.

At this early stage, it's a lot simpler to make everything load/store. Later on, when we start looking at some patterns, we can easily add values, too, and replace existing load/store patterns with them.

In the end, but still pretty soon, most patterns will be values with only actual stack variables that need their addresses taken will remain as load/store. I'd rather implement that when I have a concrete example I can test than now.

Same reason as above: move from simple to complex implementation is a lot easier than complex wrong to complex right.